instrumentation/resque/lib/opentelemetry/instrumentation/resque/patches/resque_job.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Resque
+      module Patches
+        # Module to prepend to Resque::Job for instrumentation
+        module ResqueJob
+          def perform # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+            job_class = payload_class_name
+
+            attributes = {
+              'messaging.system' => 'resque',
+              'messaging.destination' => queue.to_s,
+              'messaging.destination_kind' => 'queue',
+              'messaging.resque.job_class' => job_class
+            }
+
+            span_name = if config[:job_class_span_names]
+                          "#{job_class} process"
+                        else
+                          "#{queue} process"
+                        end
+
+            extracted_context = OpenTelemetry.propagation.extract(@payload)
+
+            links = []
+            if config[:propagation_style] == :link
+              span_context = OpenTelemetry::Trace.current_span(extracted_context).context
+              links << OpenTelemetry::Trace::Link.new(span_context) if span_context.valid?
+            end
+
+            parent_context = (config[:propagation_style] == :child ? extracted_context : OpenTelemetry::Context.current)
+
+            OpenTelemetry::Context.with_current(parent_context) do
+              tracer.in_span(span_name, attributes: attributes, links: links, kind: :consumer) do
+                super
+              end
+            end
```

Comment:
This is actually more complex to reason about than it seems. I think we need to propagate the `extracted_context` in both cases, because it may contain additional context like `baggage`, but we need to create a root span when `config[:propagation_style] == :link`. I would suggest splitting the two cases and handling them separately rather than the complex logic here that tries to use a single path for the `in_span` and `super`.

```suggestion
            OpenTelemetry::Context.with_current(extracted_context) do
              if config[:propagation_style] == :link
                links = []
                span_context = OpenTelemetry::Trace.current_span(extracted_context).context
                links << OpenTelemetry::Trace::Link.new(span_context) if span_context.valid?
                OpenTelemetry::Trace.with_span(tracer.start_root_span(span_name, attributes: attributes, links: links, kind: :consumer)) do
                  super
                end
              else
                tracer.in_span(span_name, attributes: attributes, kind: :consumer) do
                  super
                end
              end
            end
```

Reply:
Whenever I refactor code in the GitHub web UI it is obviously perfect - I'm sure you simply applied the suggestion incorrectly! :trollface:

Reply:
I'm updating #726 to match; but it occurs to me that explicitly calling `start_root_span` could have unintended consequences if a trace had already been started by something even lower-level (for instance, `redis` instrumentation).

Is that important to worry about? I'm not sure.

Reply:
We're explicitly starting a new trace for the job as a unit of work, so I think `start_root_span` makes sense here.

Additional context here: at Shopify we also trace the job worker loop (at a low sample rate) and we don't want the job trace to be associated with that "job infrastructure" trace.

Reply:
> we don't want the job trace to be associated with that "job infrastructure" trace.

I think I could support that as well, it seems reasonable.